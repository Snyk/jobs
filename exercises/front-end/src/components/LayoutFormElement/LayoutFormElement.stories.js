import '~/assets/scss/styles.scss';
import LayoutFormElement from './LayoutFormElement';

export default {
  title: 'Library/Base Components/Layout/FormElement',
  component: LayoutFormElement,
  parameters: {
    componentSubtitle:
      'The `LayoutFormElement` component ensures consistent spacing between form fields and labels.',
  },
};

export const Default = () => ({
  components: { LayoutFormElement },
  template: `
    <StoryContainer>
      <LayoutFormElement>
        <template #label>
          <label for="example">Label 1</label>
        </template>
        <input type="text" placeholder="Any kind of form element..." style="border: 1px solid #ddd; background-color: #fff; padding: 12px; width: 100%;"/>
      </LayoutFormElement>

      <LayoutFormElement>
        <template #label>
          <label for="example">Label 2</label>
        </template>
        <input type="text" placeholder="Another type of input..." style="border: 1px solid #ddd; background-color: #fff; padding: 12px; width: 100%;"/>
      </LayoutFormElement>

      <LayoutFormElement>
        <template #label>
          <label for="example">Label 3</label>
        </template>
        <input type="text" placeholder="Fill your boots..." style="border: 1px solid #ddd; background-color: #fff; padding: 12px; width: 100%;"/>
      </LayoutFormElement>
    <StoryContainer>
  `,
});

export const Regular = () => ({
  components: { LayoutFormElement },
  template: `
    <StoryContainer>
      <LayoutFormElement regular>
        <template #label>
          <label for="example">Label 1</label>
        </template>
        <input type="text" placeholder="Any kind of form element..." style="border: 1px solid #ddd; background-color: #fff; padding: 12px; width: 100%;"/>
      </LayoutFormElement>

      <LayoutFormElement regular>
        <template #label>
          <label for="example">Label 2</label>
        </template>
        <input type="text" placeholder="Another type of input..." style="border: 1px solid #ddd; background-color: #fff; padding: 12px; width: 100%;"/>
      </LayoutFormElement>

      <LayoutFormElement regular>
        <template #label>
          <label for="example">Label 3</label>
        </template>
        <input type="text" placeholder="Fill your boots..." style="border: 1px solid #ddd; background-color: #fff; padding: 12px; width: 100%;"/>
      </LayoutFormElement>
    <StoryContainer>
  `,
});
Regular.storyName = '…with regular weight';

export const Inline = () => ({
  components: { LayoutFormElement },
  template: `
    <StoryContainer>
      <LayoutFormElement inline>
        <template #label>
          <label for="example">Label 1</label>
        </template>
        <input type="text" placeholder="Any kind of form element..." style="border: 1px solid #ddd; background-color: #fff; padding: 12px; width: 100%;"/>
      </LayoutFormElement>

      <LayoutFormElement inline>
        <template #label>
          <label for="example">Label 2</label>
        </template>
        <input type="text" placeholder="Another type of input..." style="border: 1px solid #ddd; background-color: #fff; padding: 12px; width: 100%;"/>
      </LayoutFormElement>

      <LayoutFormElement inline>
        <template #label>
          <label for="example">Label 3</label>
        </template>
        <input type="text" placeholder="Fill your boots..." style="border: 1px solid #ddd; background-color: #fff; padding: 12px; width: 100%;"/>
      </LayoutFormElement>
    <StoryContainer>
  `,
});
Inline.storyName = '…as inline';

export const CustomSpacing = () => ({
  components: { LayoutFormElement },
  template: `
    <StoryContainer>
      <LayoutFormElement spacing="xs">
        <label for="example"><input type="checkbox" /> Label 1</label>
      </LayoutFormElement>
      
      <LayoutFormElement spacing="xs">
        <label for="example"><input type="checkbox" /> Label 2</label>
      </LayoutFormElement>
      
      <LayoutFormElement spacing="xs">
        <label for="example"><input type="checkbox" /> Label 3</label>
      </LayoutFormElement>
    <StoryContainer>
  `,
});
CustomSpacing.storyName = '…with custom spacing';
CustomSpacing.parameters = {
  docs: {
    storyDescription: `For elements like checkboxes and radio inputs you can customise the spacing
      between adjacent elements using the \`spacing\` prop. All global spacing sizes are
      available.`,
  },
};
