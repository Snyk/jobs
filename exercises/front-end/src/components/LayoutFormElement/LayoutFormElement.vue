<template>
  <div class="vue--layout-form-element" :class="computedClasses">
    <div v-if="showLabelSlot" class="vue--layout-form-element__label">
      <slot name="label" />
    </div>
    <div class="vue--layout-form-element__field">
      <slot />
    </div>
  </div>
</template>

<script>
export default {
  name: 'LayoutFormElement',
  status: 'beta',

  props: {
    /**
     * Whether the form input field is displayed inline.
     */
    inline: {
      type: Boolean,
      default: false,
    },
    /**
     * Should fill the available width.
     */
    fillWidth: {
      type: Boolean,
      default: false,
    },
    /**
     * Set the adjacent component spacing.
     */
    spacing: {
      type: String,
      default: null,
    },
    /**
     * Regular weight label.
     */
    regular: {
      type: Boolean,
      default: false,
    },
  },

  computed: {
    computedClasses() {
      return {
        [`vue--layout-form-element--${this.spacing}`]: this.spacing,
        'vue--layout-form-element--inline': this.inline,
        'vue--layout-form-element--fill-width': this.fillWidth,
        'vue--layout-form-element--regular': this.regular,
      };
    },
    showLabelSlot() {
      return !!this.$slots['label'];
    },
  },
};
</script>

<style lang="scss" scoped>
@import '~/assets/scss/utils';

.vue--layout-form-element {
  $self: &;

  + #{$self} {
    margin-top: space();
  }

  &__label {
    display: block;
    line-height: 1.3;
    margin: 0 space(xs) space(xs) 0;
  }

  &:not(&--regular) &__label {
    @include font-stack(semibold);
  }

  &__field {
    display: block;
  }

  &--fill-width {
    #{$self}__field {
      width: 100%;
    }
  }

  &--inline {
    align-items: center;
    display: flex;

    #{$self}__label {
      margin-bottom: 0;
      margin-right: space(s);
    }

    &#{$self}--fill-width {
      #{$self}__field {
        width: 100%;
      }
    }
  }

  &--none + #{$self} {
    margin-top: 0;
  }

  @each $key, $value in map-get($spacing, 'default') {
    &--#{$key} + #{$self} {
      margin-top: space($key);
    }
  }
}
</style>
