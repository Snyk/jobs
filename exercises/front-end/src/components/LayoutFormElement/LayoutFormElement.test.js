import { shallowMount } from '@vue/test-utils';
import LayoutFormElement from '~/components/LayoutFormElement/LayoutFormElement';

describe('LayoutFormElement', () => {
  describe('Accessibility', () => {
    it('passes basic accessibility checks', async () => {
      const wrapper = shallowMount(LayoutFormElement);

      await expect(wrapper).toBeAccessible();
    });
  });

  describe('Snapshots', () => {
    it('renders as expected', () => {
      const wrapper = shallowMount(LayoutFormElement, {
        slots: {
          label: 'BaseBadge goes here',
          default: 'Field goes here',
        },
      });

      expect(wrapper).toMatchSnapshot();
    });

    it('renders without label markup', () => {
      const wrapper = shallowMount(LayoutFormElement, {
        slots: {
          default: 'Field goes here',
        },
      });

      expect(wrapper).toMatchSnapshot();
    });

    it('renders valid html', () => {
      const wrapper = shallowMount(LayoutFormElement, {
        slots: {
          default: 'Field goes here',
        },
      });

      expect(wrapper.html()).toHTMLValidate();
    });
  });
});
