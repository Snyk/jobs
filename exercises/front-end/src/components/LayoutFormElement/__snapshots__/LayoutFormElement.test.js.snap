// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LayoutFormElement Snapshots renders as expected 1`] = `
<div class="vue--layout-form-element">
  <div class="vue--layout-form-element__label">BaseBadge goes here</div>
  <div class="vue--layout-form-element__field">Field goes here</div>
</div>
`;

exports[`LayoutFormElement Snapshots renders without label markup 1`] = `
<div class="vue--layout-form-element">
  <!---->
  <div class="vue--layout-form-element__field">Field goes here</div>
</div>
`;
